system/datatypes/mailinglist.php

```diff
@@ -697,14 +704,60 @@ protected function mailingView(HyphaRequest $request) {
-			$addresses = $this->findAddresses($xpath);
-			$num = count($addresses);
-
+			$num = $this->getNumberOfAddresses();
 			$linkToMailing = $this->path(self::PATH_MAILS_VIEW_ID, ['id' => $mailingId]);
-			$action = 'if(confirm(' . json_encode(__('ml-sure-to-send', ['count' => $num])) . '))' . makeAction($linkToMailing, self::CMD_SEND, '');
-			$action = str_replace(['"'], ['\''], $action);
-			$commands->append(makeButton(__('send'), $action));
+
+			$action = 'if(confirm(' . json_encode(__('ml-sure-to-send', ['count' => $num])) . ')){
+				const divPopUp = document.createElement("DIV"); // create a popup
+				divPopUp.className = "messagebox modal";
+				const divProgress = document.createElement("div");
+				divProgress.className = "progress";
+
+				const spanProgressText = document.createElement("span");
+				spanProgressText.innerHTML= "De mail wordt verstuurd...";
+				spanProgressText.className = "progress-text";
+				const spanProgressIndicator = document.createElement("span");
+				spanProgressIndicator.className = "progress-indicator";
+				spanProgressIndicator.innerHTML = "0/'.$num.'";
+				const divWarning = document.createElement("div");
+				divWarning.className = "warning";
+				divWarning.innerHTML = "Let op: Sluit dit venster niet tot de hele mailing is verstuurd";
+				divProgress.appendChild(spanProgressText);
+				divProgress.appendChild(spanProgressIndicator);
+				divPopUp.appendChild(divProgress);
+				divPopUp.appendChild(divWarning);
+
+				const divGray = document.createElement(\'DIV\'); // create a gray layer
+				divGray.className="gray-modal-background";
+				document.body.appendChild(divGray);
+				document.body.appendChild(divPopUp);
+				let checkProgressInterval;
+				const onCompleted = function(){
+					if(checkProgressInterval) clearInterval(checkProgressInterval);
+					checkProgressInterval = null;
+					if(divPopUp && divPopUp.parentNode === document.body) document.body.removeChild(divPopUp);
+					if(divGray && divGray.parentNode === document.body) document.body.removeChild(divGray);
+					window.onbeforeunload = undefined;
+				};
+				const updateProgress = function(progress){
+					const [done,total] = progress.split("/").map(Number);
+					if(done>=total) onCompleted();
+					else{
+						spanProgressIndicator.innerHTML = progress;
+					}
+				};
+			  const checkProgress = function(){'. // retrieve the progress of the send action
+					makeAction($linkToMailing, self::CMD_PROGRESS, '', null, 'updateProgress')
+				.'};
+				checkProgressInterval = setInterval(checkProgress,1000); //TODO make into a constant
+				window.onbeforeunload = function(){return \'Are you sure you want to leave?\';};' // lock the browser from exiting. TODO translation?
+				 . makeAction($linkToMailing, self::CMD_SEND, '', null, 'onCompleted').';
+			 }';
+
+			if ($num > 0){ // hide send button if no addresses available
+				$action = str_replace(['"'], ['\''], $action);
```

Comment:
Why is this replace needed? If this is because `"` breaks the `onclick="..."` quoting, then maybe this should be fixed by using `htmlspecialchars` in `makeAction` instead.

Reply:
Seems this is from existing code, but still interesting to look into this.

Reply:
Yeah, I just kept it there. I can look into it if that's helpful?

Reply:
I suspect there is some fragility in the current system because of this, so if you want to have look, please do.

Reply:
```
system/core/various.php:98

function makeButton($label, $action, $id = '', $class = '') {
		$class = 'button' . ($class ? ' ' . $class : '');
		return '<input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'.$action.'" />';
	}
```

I would suggest doing a fix here instead of in mailinglist.php. We could do something along these lines:

```
function makeButton($label, $action, $id = '', $class = '') {
		$class = 'button' . ($class ? ' ' . $class : '');
		$action = makeSave($action);
		return '<input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'.$action.'" />';
	}
```

but I wouldn't recommend using `htmlspecialchars`. Because that will make html entities out of javascript code, right?
Simply replacing `"` with `\'` could also lead to unpredictable behavior, depending on your javascript usage of quotes. Same with replacing `"` with `\"`. There's always an edge case that could be triggered.

Best way I can think of might be:

```
function makeButton($label, $action, $id = '', $class = '') {
		$class = 'button' . ($class ? ' ' . $class : '');
		$functionName = '?';
		return '<script>function '.$functionName.'(){'.$action.'}</script><input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'$functionName'();" />';
	}
```

But then we would need a proper way of determining a good unique name for `$functionName`;

Reply:
> but I wouldn't recommend using htmlspecialchars. Because that will make html entities out of javascript code, right?

It would, but I don't think that's a problem. Since HTML attributes  (such as onclick) should be entity-decoded before they are used (this should be handled transparently by the DOM parser), so the javascript interpreter should see exactly the input the `htmlspecialchars`. One notable exception is CDATA (inside CDATA blocks or some tags), which is used literal, but that doesn't apply here AFAICS.

> Best way I can think of might be:

Clever approach, but it also seems a bit fragile. And indeed, it needs unique function names.

Reply:
htmlspecialchars it is :-)

Reply:
Though this does leave the edge case that, for example, `console.log('&gt;')` will not perform as expected.

Reply:
Why not? Did you test this? I think this should be encoded as `console.log(&amp;gt;')` in the HTML, which should be decoded into the proper code again before running the javascript?

Reply:
My bad, apparently I tested it to hastily and messed up enabling the `htmlspecialchars` :-S